src/text/writer.rs

```diff
@@ -0,0 +1,352 @@
+use crate::result::{illegal_operation, IonResult};
+use crate::text::encoder::TextEncoder;
+use crate::IonType;
+use bigdecimal::BigDecimal;
+use chrono::{DateTime, FixedOffset};
+use std::io::{BufWriter, Write};
+
+pub struct TextWriter<W: Write> {
+    output: BufWriter<W>,
+    annotations: Vec<String>,
+    field_name: Option<String>,
+    containers: Vec<IonType>,
+}
+
+impl<W: Write> TextWriter<W> {
+    /// Constructs a new instance of TextWriter that writes values to the provided io::Write
+    /// implementation.
+    pub fn new(sink: W) -> TextWriter<W> {
+        TextWriter {
+            output: BufWriter::new(sink),
+            annotations: vec![],
+            field_name: None,
+            containers: vec![],
+        }
+    }
+
+    /// Returns a reference to the underlying io::Write implementation.
+    pub fn output(&self) -> &W {
+        self.output.get_ref()
+    }
```

Comment:
Hmm, is this a useful API to expose?  What is the use case for it (out of curiosity)?

Reply:
It's an API pattern I've encountered; things that own a `io::Write` tend to provide a reference to it and a way to retrieve it when the wrapper is no longer needed. See `BufWriter`'s [`get_ref`](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.get_ref) and [`into_inner`](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_inner) methods for an example.

It provides a nice way to do read-only queries against the underlying value (like getting [the peer IP address](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.peer_addr) from a `TcpStream` struct) and then ultimately do something else with it when you're done (like reading a TCP response back after writing the request).

I haven't added an `into_inner` method yet, but I think it's a good idea.

Reply:
The use case around `into_write` or some such makes sense to me (assuming your at a place that it makes sense to return the underlying `Write`), breaking encapsulation to get at the underlying `Write` by reference seems a little weak (kind of makes sense for a generic API like `BufWriter`, but less clear about what I can do in any reasonable way around an Ion writer.

It's not a big deal, but I always like to have less API unless we have good reasons to expose these underlying bits.